src/librustc_macros/src/session_diagnostic.rs

```diff
@@ -0,0 +1,622 @@
+            syn::Meta::List(list) => {
+                match list.path.segments.iter().last().unwrap().ident.to_string().as_str() {
+                    suggestion_kind @ "suggestion"
+                    | suggestion_kind @ "suggestion_short"
+                    | suggestion_kind @ "suggestion_hidden"
+                    | suggestion_kind @ "suggestion_verbose" => {
+                        // For suggest, we need to ensure we are running on a (Span,
+                        // Applicability) pair.
+                        let (span, applicability) = (|| {
+                            if let syn::Type::Tuple(tup) = &info.ty {
+                                let mut span_idx = None;
+                                let mut applicability_idx = None;
+                                for (idx, elem) in tup.elems.iter().enumerate() {
+                                    if type_matches_path(elem, &["rustc_span", "Span"]) {
+                                        if span_idx.is_none() {
+                                            span_idx = Some(syn::Index::from(idx));
+                                        } else {
+                                            throw_span_err!(
+                                                info.span.clone().unwrap(),
+                                                "type of field annotated with `#[suggestion(...)]` contains more than one Span"
+                                            );
+                                        }
+                                    } else if type_matches_path(
+                                        elem,
+                                        &["rustc_errors", "Applicability"],
+                                    ) {
+                                        if applicability_idx.is_none() {
+                                            applicability_idx = Some(syn::Index::from(idx));
+                                        } else {
+                                            throw_span_err!(
+                                                info.span.clone().unwrap(),
+                                                "type of field annotated with `#[suggestion(...)]` contains more than one Applicability"
+                                            );
+                                        }
+                                    }
+                                }
+                                if let (Some(span_idx), Some(applicability_idx)) =
+                                    (span_idx, applicability_idx)
+                                {
+                                    let binding = &info.binding.binding;
+                                    let span = quote!(#binding.#span_idx);
+                                    let applicability = quote!(#binding.#applicability_idx);
+                                    return Ok((span, applicability));
+                                }
+                            }
+                            throw_span_err!(
```

Comment:
We could just expect a specific trait to be implemented for these fields (and give that trait an `on_unimplemented` attribute for a targetted diagnostic). Then you can invoke `ThatTrait::get_span(#binding)` and `ThatTrait::get_applicability`. `ThatTrait` could be implemented for various combinations of `Span` and `Applicability` or the lack of the latter.

Reply:
I'm all in favour of removing the name-based type checks in favour of just letting custom traits do the heavy lifting. However, doing this, the emitted error points at the `#[derive(SessionDiagnostic)]`, even with `quote_spanned!` (as with the other case):

https://github.com/jumbatm/rust/blob/0254a1831fd926207b51bc87cd9deb3de7e5ee5e/src/test/ui-fulldeps/session-derive-errors.stderr#L164-L171

On having an `on_unimplemented` method: wouldn't I need to generate code which calls this method, meaning I'd be emitting the customised diagnostic at runtime? Using `compile_error!` wouldn't do the trick, either, that would always cause a compile error (regardless of the flow at runtime).